ui-lib/src/main/java/co/electriccoin/zcash/ui/common/provider/AddressBookProvider.kt

```diff
@@ -24,18 +33,71 @@ class AddressBookProviderImpl : AddressBookProvider {
         addressBook: AddressBook
     ) {
         file.outputStream().use {
-            serializeAddressBookToByteArrayFile(it, addressBook)
+            serializeAddressBook(it, addressBook)
         }
     }
 
     override fun readAddressBookFromFile(file: File): AddressBook {
         return file.inputStream().use {
-            deserializeByteArrayFileToAddressBook(it)
+            deserializeAddressBook(it)
         }
     }
 
-    private fun serializeAddressBookToByteArrayFile(
-        outputStream: FileOutputStream,
+    /**
+     * Serializes the given address book and then encrypts it for the given account.
+     *
+     * The ciphertext is written to the provided output stream.
+     */
+    private fun encryptAddressBook(
+        addressBookKey: AddressBookKey,
+        outputStream: OutputStream,
+        addressBook: AddressBook
+    ) {
+        // Generate a fresh one-time key for this ciphertext.
+        val salt = Random.randBytes(32)
+        val key = addressBookKey.deriveEncryptionKey(salt)
+
+        val plaintext = ByteArrayOutputStream().use { stream ->
+            serializeAddressBook(stream, addressBook)
+            stream.toByteArray()
+        }
+
+        // Tink encodes the ciphertext as `nonce || ciphertext || tag`.
+        val cipher = ChaCha20Poly1305.create(key)
+        val ciphertext = cipher.encrypt(plaintext, null)
+
+        outputStream.buffered().use {
+            it.write(salt)
+            it.write(ciphertext)
+        }
+    }
```

Comment:
Can this please have a version byte prepended to it? Or is the version going to be stored somewhere else?

Reply:
There's a version byte in the plaintext, and my original understanding was that the ciphertext was going to be stored in a way that handled external versioning. But the wallet side has changed somewhat and I no longer know how the file is intended to be managed from that perspective.

Reply:
There are two ways of doing versioning for the encryption, as opposed to the encrypted data:
a) derive keys with a domain separator (currently `"ZashiAddressBookEncryptionV1"`) and trial-decrypt with each such separator;
b) prepend a separate encryption version byte;
c) rely on external context.

I dislike c) but have no strong opinion between a) and b).

Reply:
we currently do b)